libraries/clchain/include/clchain/graphql.hpp

```diff
@@ -0,0 +1,769 @@
+                  }
+                  else
+                  {
+                     using mf = eosio::member_fn<member_type>;
+                     eosio::tuple_from_type_list<typename mf::arg_types> args;
+                     bool filled[mf::num_args] = {};
+                     gql_mark_optional<0>(args, filled);
+                     if (input_stream.current_puncuator == '(')
+                     {
+                        input_stream.skip();
+                        while (input_stream.current_type == gql_stream::name)
+                        {
+                           bool found = false;
+                           if (!gql_parse_args<0>(args, filled, found, input_stream, error,
+                                                  arg_names...))
+                              return (ok = false), void();
+                           if (!found)
+                              return (ok = error("unknown arg '" +
+                                                 (std::string)input_stream.current_value + "'")),
+                                     void();
+                        }
+                        if (input_stream.current_puncuator != ')')
+                           return (ok = error("expected )")), void();
+                        input_stream.skip();
+                     }
+                     if constexpr (mf::num_args > 0)
+                        for (int i = 0; i < mf::num_args; ++i)
+                           if (!filled[i])
+                              return (ok = error("function missing required arg '" +
+                                                 std::string(std::data({arg_names...})[i]) + "'")),
+                                     void();
+                     auto result = std::apply(
+                         [&](auto&&... args) {
+                            return (value.*member(&value))(std::move(args)...);
+                         },
+                         args);
+                     if (!gql_query(result, input_stream, output_stream, error))
+                        return (ok = false), void();
+                  }
+               }
+            }
+         });
+         if (!ok)
+            return false;
+         if (!found)
+            return error((std::string)field_name + " not found");
+      }
+      if (input_stream.current_puncuator != '}')
+         return error("expected }");
+      input_stream.skip();
+      if (!first)
+      {
+         decrease_indent(output_stream);
+         write_newline(output_stream);
+      }
+      output_stream.write('}');
+      return true;
+   }
+
+   template <typename T, typename OS, typename E>
+   bool gql_query_root(const T& value, gql_stream& input_stream, OS& output_stream, const E& error)
+   {
+      if (input_stream.current_type == gql_stream::name)
+      {
+         if (input_stream.current_value == "query")
+            input_stream.skip();
+         else
+            return error("expected query");
+      }
```

Comment:
Added explicit unsupported errors

Reply:
Allowing the `Name` should be easy—it's only relevant to distinguish which operation to execute, and can therefore be ignored.  For the rest, I'm concerned that leaving them unimplemented is not very future-proof.  Variables and multiple operations both require additional input, so adding support for them them is API breaking.  Supporting fragments would require a completely different implementation, because a selection set can reference a fragment defined later in the document.